src/lwrap.1
.TH lwrap 1
.SH NAME
lwrap
.SH SYNOPSIS
.B lwrap
[-c columns] [-j columns][-e][-f text file]
.SH DESCRIPTION
Line wrap text input according to user defined line width.
.SH EXECUTION ENVIRONMENT
LANG=countrycode.utf8
.SH OPTIONS
.TP
.I "\-h"
Print usage and exit.
.TP
.I "\-c"
Set number of columns before line break.
The default is 80 (min:1,max:235).
Each three bytes glyph (e.g. CJK) takes 2 columns.
Each one or two bytes glyph (e.g. ascii and latin)
takes 1 column.
.TP
.I "\-e"
Line break without ascii space but a newline only.
The default line break is ascii space plus newline, 
it makes the output reversible back to it's original 
input format. The reason is ascii space occasionally 
been used as word and name separator in asian 
languages, and if these spaces were erased, the output 
will no longer be reversible. The default can be 
changed with this command option, which may be the  
desired final output.

.TP
.I "\-j"
Same as -c but justified lines with ascii spaces.
.TP
.I "\-f"
read input from text file instead of default stdin.
.TP
.I "\-d"
Output debugging text to /tmp/lwrap.debug file.
.TP
.I "\-V"
Print Version and exit.
.SH USE WITH VIM
.VE
.LP
    :[range]!lwrap -j80
    or inside file VIMRC
    set formatprg=lwrap\ -j80\

.SH AUTHORS
written by Yang Li.
.SH REPORTING BUGS
github.com/netcrop/lwrap
.SH LIMITATIONS
Many natural languages use different glyph width within the same font,
the output line width may not be consistent in these languages.
.SH LICENSE
GNU GENERAL PUBLIC LICENSE Version 2

.SH EXAMPLES
.VE
.LP
  The  franchise  began  in 1977 	実写映画本編
  with  the  release of the film 	本シリーズの実写映画本編は、複
  Star   Wars  (later  subtitled 	数のエピソードにより綴られるサ
  Episode  IV:  A  New  Hope  in 	ーガの形式を採っている。エピソ
  1981),[2][3]  which  became  a 	ード番号は、原題ではローマ数字
  worldwide      pop     culture 	を採用しているが、邦題ではアラ
  phenomenon. It was followed by 	ビア数字を採用している。
  the   successful  sequels  The 	映画が『エピソード1/ファントム
  Empire Strikes Back (1980) and 	・メナス』からではなく『エピソ
  Return  of  the  Jedi  (1983); 	ード4/新たなる希望』から制作さ
  these  three  films constitute 	れたのは、まずは一作目が商業的
  the    original    Star   Wars 	に成果を収めねばシリーズ化が望
  trilogy. A prequel trilogy was 	めず、更に一番“冒険活劇”とし
  released   between   1999  and 	ての完成度の高かった『エピソー
  2005,   which  received  mixed 	ド4/新たなる希望』を最初に世に
  reactions  from  both  critics 	出すことが得策だったためである
  and  fans.  A  sequel  trilogy 	。また『エピソード1・2・3』の
  began in 2015 with the release 	時代は、全銀河の首都である大都
  of   Star   Wars:   The  Force 	市惑星コルサントの描写や、銀河
  Awakens.  All seven films were 	共和国と独立星系連合の間で勃発

  De  franchise portretteert een 	《星球大战》（英语：Star  Wars
  universum dat zich afspeelt in 	），中文可簡稱「星戰」，是美國
  een  sterrenstelsel  dat wordt 	導演兼編劇乔治·卢卡斯所構思拍
  omschreven  als  ver, ver weg. 	攝的一系列科幻電影。同时“星球
  Het toont vaak de Jedi als een 	大戰”也是该系列電影最早拍摄上
  representatie  van  het goede, 	映的第四章故事的原始片名。
  in  strijd  met  de  Sith, hun 	从1970年代末到1980年代，盧卡斯
  kwade  tegenhanger.  Hun wapen 	出品了《星際大戰》三部曲。他曾
  van  keuze,  het  lichtzwaard, 	説明第一部星戰（第四部曲）的人
  wordt  algemeen  erkend  in de 	物及故事是參照日本導演黑澤明的
  popcultuur.    Het    fictieve 	《戰國英豪》（隠し砦の三悪人）
  universum   bevat   ook   veel 	所創作而成；原三部曲以二戰英德
  thema's,  met  name  invloeden 	大戰為架構並溶入美國西部片及日
  van  filosofie en religie. Een 	本武士刀劇的節奏；藴涵了太空冒
  voorbeeld hiervan is De Kracht 	險、希臘神話的大視野制作。影片

  Star  Wars (à l'origine nommée 	《스타워즈》(Star  Wars),  재개
  sous  son  titre  français, La 	봉 제목 《스타워즈 에피소드 4: 
  Guerre  des  étoiles)  est  un 	새로운  희망》(Star  Wars 
  univers   fantastique   et  de 	Episode  IV:  A New Hope)은 1977
  science-fiction    créé    par 	년  미국에서  제작된 SF 에픽 스
  George  Lucas.  D'abord conçue 	페이스  오페라  영화이다. 조지 
  comme       une       trilogie 	루카스가 감독과 각본을 맡았다. 
  cinématographique sortie entre 	《스타  워즈》  오리지널 삼부작 
  1977    et   1983,   la   saga 	중  가장 처음 만들어진 작품이며
  s'élargit   ensuite  de  trois 	,  시나리오  상으로는 4번째 에피
  nouveaux  films  entre 1999 et 	소드에  해당한다.  마크 해밀, 해
  2005,  narrant  des événements 	리슨  포드,  캐리 피셔, 피터 커
  antérieurs   à   la   première 	싱,  앨릭  기니스, 데이비드 프로
  trilogie.    Cette    dernière 	스,  제임스  얼 존스, 앤서니 대
  (épisodes  IV,  V et VI) ainsi 	니얼스,  케니  베이커, 피터 메이


src/lwrap.c
/*
 * Yang Li
 * GPLv2
 */

#include "lwrap.h"

inline void flinebreak()
{
    if (my.hcolcount <= my.hcolsize)
        return;
    if (my.currdata->val == EOS)
        return;
    if ((my.linebreak = my.currdata->dataindex - my.currdata->wbytecount) < 0)
        my.linebreak += my.databoundry;
    my.lbdata = &my.data[my.linebreak];
    my.hcolcount = my.currdata->wcolcount;
    my.hbytecount = my.currdata->wbytecount;
    my.funicode[my.lbdata->val]->linebreak();
    if (my.currjustify->remain != 0 && my.currjustify->wcount == 0) ;
    else
        my.currjustify = my.currjustify->after;
}

void usage()
{
    printf("%s%s%s", " Usage: ", my.progname,
           " [-c columns] [-j columns][-e][-f text file]\n\
      LANG = countrycode.utf8\n\
      -c:   set number of columns before line break.\n\
      -j:   same as -c but justified lines with ASCII spaces.\n\
      -f:   read input from text file instead of default stdin.\n\
      -e:   line break without ASCII space but a newline only.\n\
      -h:   print usage and exit.\n\
      -d:   output debugging text to /tmp/lwrap.debug file.\n\
      -V:   print version and exit.\n\
      e.g.  -c80 -f foo.txt\n\
      e.g.  -j60 <foo.txt\n\
      e.g.  -e -j60 <foo.txt >result.txt\n\
      use with vim.  :[range]! -j60\n");
}

void version()
{
    printf("%s %s %s\n", VERSION, __TIME__, __DATE__);
}

void delocate()
{
    if (!me)
        return;
    if (my.filep != NULL){
        fclose(my.filep);
    }
    if (my.filedebug != NULL){
        fclose(my.filedebug);
    }
    if (my.data != NULL){
        free(my.data);
    }
    if (my.outindex != NULL)
        free(my.outindex);
    if (my.readindex != NULL)
        free(my.readindex);
    if (my.writebuff != NULL)
        free(my.writebuff);
    if (my.readbuff != NULL)
        free(my.readbuff);
    if (my.justify != NULL)
        free(my.justify);
    if (my.string != NULL)
        free(my.string);
    if (my.funicode != NULL)
        free(my.funicode);
    if (my.call != NULL)
        free(my.call);
    if (my.fring != NULL)
        free(my.fring);
    free(me);
    me = NULL;
}

void lwrap_mlocate(void **p, size_t size)
{
    if (!(*p = malloc(size))) {
        fprintf(stderr, "%s\n", "lwrap_mlocate: memory allocation faild.");
        exit(0);
    }
}

inline void strappend(char *str, char **strp)
{
    *strp = my.currstring;
    while ((*my.currstring++ = *str++) != EOS) ;
}

void option()
{
    while ((my.optchr = getopt(my.argc, my.argv, "hc:dj:ef:v")) != EOF) {
        switch (my.optchr) {
        case 'h':
        case '?':
            my.usage();
            exit(0);
        case 'c':
            if ((my.hcolsize =
                 (int)strtol(optarg, &my.strtolend, 10)) > 0
                && my.hcolsize < my.displaymax) {
                my.writebuffrelay = my.hcolsize * my.justifymultiple;
                break;
            }
            exit(0);
        case 'd':
            my.debug = 1;
            if ((my.filedebug = fopen(my.debugfile, "w")) == NULL)
                my.filedebug = stderr;
            break;
        case 'e':
            my.endingspace = 0;
            break;
        case 'j':
            if ((my.hcolsize =
                 (int)strtol(optarg, &my.strtolend, 10)) > 0
                && my.hcolsize < my.displaymax) {
                my.writebuffrelay = my.hcolsize * my.justifymultiple;
                my.justifying = 1;
                break;
            }
            exit(0);
        case 'f':
            if ((my.filep = fopen(optarg, "r")) != NULL)
                break;
            exit(0);
        case 'v':
            version();
            exit(0);
        }
    }
}
inline void sighandler(int sig)
{
    exit(0);
}
inline void avariable()
{
    atexit(delocate);
    signal(SIGINT,sighandler);
    if(me)
        return;
    lwrap_mlocate((void **)&me, sizeof(self));
    my.displaymax = 236;
    my.justifymultiple = 2;
    my.buffmultiple = 42;
    my.argc = 0;
    my.argv = (char **)NULL;
    my.foutput = &foutput;
    my.usage = &usage;
    my.version = &version;
    my.option = &option;
    my.freadfile = &freadfile;
    my.delocate = &delocate;
    my.fjustify = &fjustify;
    my.alocate = &alocate;
    my.flinebreak = &flinebreak;
    my.hcolsize = 80;
    my.maxwordsize = 28;
    my.readbuffrelay = 2;
    my.writebuffrelay = my.maxwordsize;
    my.callsize = 2;
    my.flinebreaksize = 28;
    my.fmiddlebytesize = 28;
    my.fringboundry = 28;
    my.fringsize = my.fringboundry - 1;
    my.initwritebuff = FINITOUTBUFF;
    my.buffboundry = my.hcolsize * my.buffmultiple + 1;
    my.readbuffsize = my.buffboundry - 1;
    my.databoundry = my.hcolsize * my.buffmultiple + 1;
    my.datasize = my.databoundry - 1;
    my.writebuffboundry = my.hcolsize * my.buffmultiple + 1;
    my.writebuffsize = my.writebuffboundry - 1;
    my.writesize = 3200;
    my.justifyboundry = 40;
    my.justifysize = my.justifyboundry - 1;
    my.hcolcount = 0;
    my.hbytecount = 0;
    my.linebreak = 0;
    my.getchar = 0;
    my.wcolcount = 0;
    my.delimeter = 0;
    my.justifying = 0;
    my.endingspace = 1;
    my.wbytecount = 0;
    my.optchr = 0;
    my.debug = 0;
    my.readsize = 0;
    my.rindex = 0;
    my.oindex = 0;
    my.stringsize = 60;
    my.filesize = 0;
    my.strtolend = (char *)NULL;
    my.filep = stdin;
    my.filedebug = stderr;
    my.data = (data_t *) NULL;
    my.outdata = (data_t *) NULL;
    my.currdata = (data_t *) NULL;
    my.prevdata = (data_t *) NULL;
    my.nextdata = (data_t *) NULL;
    my.readdata = (data_t *) NULL;
    my.lbdata = (data_t *) NULL;
    my.outindex = (index_t *) NULL;
    my.readindex = (index_t *) NULL;
    my.call = (fun) NULL;
    my.justify = (justify_t *) NULL;
    my.funicode = (fun_t **) NULL;
    my.readbuff = (char *)NULL;
    my.writebuff = (char *)NULL;
    my.fring = (fun_t **) NULL;
    lwrap_mlocate((void **)&my.string, (sizeof(char) * my.stringsize));
    for (int i = 0; i < my.stringsize; i++)
        my.string[i] = SPACE;
    my.string[my.stringsize - 1] = EOS;
    my.currstring = &my.string[0];
    strappend("lwrap", &my.progname);
    strappend("/tmp/lwrap.debug", &my.debugfile);
}

inline void acall()
{
    lwrap_mlocate((void **)&my.call, (sizeof(fun) * my.callsize));
    my.call[0] = &finitwritebuff;
    my.call[1] = &fwritebuff;
}

inline void afring()
{
    lwrap_mlocate((void **)&my.fring, (sizeof(fun_t) * my.fringboundry));
    my.fring[0].filter = &nonealnum;
    my.fring[0].linebreak = &lbnonealnum;
    my.fring[1].filter = &alnum;
    my.fring[1].linebreak = &lbalnum;
    my.fring[2].filter = &space;
    my.fring[2].linebreak = &lbspace;
    my.fring[3].filter = &newline;
    my.fring[3].linebreak = &lbnewline;
    my.fring[4].filter = &formfeed;
    my.fring[4].linebreak = &lbformfeed;
    my.fring[5].filter = &middlebyte;
    my.fring[5].linebreak = &lbmiddlebyte;
    my.fring[6].filter = &invalidbyte;
    my.fring[6].linebreak = &lbinvalidbyte;
    my.fring[7].filter = &latinheader;
    my.fring[7].linebreak = &lblatinheader;
    my.fring[8].filter = &ipaheader;
    my.fring[8].linebreak = &lbipaheader;
    my.fring[9].filter = &accentsheader;
    my.fring[9].linebreak = &lbaccentsheader;
    my.fring[10].filter = &greekheader;
    my.fring[10].linebreak = &lbgreekheader;
    my.fring[11].filter = &cyrilheader;
    my.fring[11].linebreak = &lbcyrilheader;
    my.fring[12].filter = &armeniheader;
    my.fring[12].linebreak = &lbarmeniheader;
    my.fring[13].filter = &hebrewheader;
    my.fring[13].linebreak = &lbhebrewheader;
    my.fring[14].filter = &arabicheader;
    my.fring[14].linebreak = &lbarabicheader;
    my.fring[15].filter = &syriacheader;
    my.fring[15].linebreak = &lbsyriacheader;
    my.fring[16].filter = &thaanaheader;
    my.fring[16].linebreak = &lbthaanaheader;
    my.fring[17].filter = &nkoheader;
    my.fring[17].linebreak = &lbnkoheader;
    my.fring[18].filter = &indicheader;
    my.fring[18].linebreak = &lbindicheader;
    my.fring[19].filter = &mischeader;
    my.fring[19].linebreak = &lbmischeader;
    my.fring[20].filter = &symbolheader;
    my.fring[20].linebreak = &lbsymbolheader;
    my.fring[21].filter = &cjkheader;
    my.fring[21].linebreak = &lbcjkheader;
    my.fring[22].filter = &asianheader;
    my.fring[22].linebreak = &lbasianheader;
    my.fring[23].filter = &puaheader;
    my.fring[23].linebreak = &lbpuaheader;
    my.fring[24].filter = &formsheader;
    my.fring[24].linebreak = &lbformsheader;
    my.fring[25].filter = &hangulheader;
    my.fring[25].linebreak = &lbhangulheader;
    my.fring[26].filter = &fourbyteheader;
    my.fring[26].linebreak = &lbfourbyteheader;
}

inline void aunicode()
{
    lwrap_mlocate((void **)&my.funicode, (sizeof(fun_t *) * UNICODESIZE));
    for (int i = 0; i <= NONEALNUMMAX; i++)
        my.funicode[i] = &my.fring[0];
    for (int i = ALNUMMIN; i <= ALNUMMAX; i++)
        my.funicode[i] = &my.fring[1];
    my.funicode[ALNUMMAX + 1] = &my.fring[0];
    my.funicode[SPACE] = &my.fring[2];
    my.funicode[NEWLINE] = &my.fring[3];
    my.funicode[FORMFEED] = &my.fring[4];
    for (int i = MIDDLEBYTEMIN; i <= MIDDLEBYTEMAX; i++)
        my.funicode[i] = &my.fring[5];
    my.funicode[192] = &my.fring[6];
    my.funicode[193] = &my.fring[6];
    for (int i = LATINMIN; i <= LATINMAX; i++)
        my.funicode[i] = &my.fring[7];
    for (int i = IPAMIN; i <= IPAMAX; i++)
        my.funicode[i] = &my.fring[8];
    for (int i = ACCENTSMIN; i <= ACCENTSMAX; i++)
        my.funicode[i] = &my.fring[9];
    for (int i = GREEKMIN; i <= GREEKMAX; i++)
        my.funicode[i] = &my.fring[10];
    for (int i = CYRILMIN; i <= CYRILMAX; i++)
        my.funicode[i] = &my.fring[11];
    my.funicode[ARMENIMIN] = &my.fring[12];
    for (int i = HEBREWMIN; i <= HEBREWMAX; i++)
        my.funicode[i] = &my.fring[13];
    for (int i = ARABICMIN; i <= ARABICMAX; i++)
        my.funicode[i] = &my.fring[14];
    my.funicode[SYRIACMIN] = &my.fring[15];
    my.funicode[THAANAMIN] = &my.fring[16];
    my.funicode[NKOMIN] = &my.fring[17];
    my.funicode[INDICMIN] = &my.fring[18];
    my.funicode[MISCMIN] = &my.fring[19];
    my.funicode[SYMBOLMIN] = &my.fring[20];
    for (int i = CJKMIN; i <= CJKMAX; i++)
        my.funicode[i] = &my.fring[21];
    my.funicode[ASIANMIN] = &my.fring[22];
    my.funicode[PUAMIN] = &my.fring[23];
    my.funicode[FORMSMIN] = &my.fring[24];
    for (int i = HANGULMIN; i <= HANGULMAX; i++)
        my.funicode[i] = &my.fring[25];
    for (int i = THREEBYTEHEADER; i < FOURBYTEHEADER; i++)
        my.funicode[i] = &my.fring[26];
    for (int i = FOURBYTEHEADER; i < UNICODESIZE; i++)
        my.funicode[i] = &my.fring[6];
}

inline void abuff()
{
    lwrap_mlocate((void **)&my.readbuff, (sizeof(char) * my.buffboundry));
    for (int i = 0; i < my.buffboundry; i++)
        my.readbuff[i] = EOS;
    lwrap_mlocate((void **)&my.writebuff, (sizeof(char) * my.writebuffboundry));
    for (int i = 0; i < my.writebuffboundry; i++)
        my.writebuff[i] = EOS;
    lwrap_mlocate((void **)&my.outindex, (sizeof(index_t) * my.writebuffboundry));
    my.outindex[0].val = 0;
    my.outindex[0].decre = &my.outindex[my.writebuffsize];
    my.outindex[my.writebuffsize].incre = &my.outindex[0];
    my.curroutindex = &my.outindex[0];
    for (int i = 1; i < my.writebuffboundry; i++) {
        my.outindex[i].val = i;
        my.outindex[i].decre = &my.outindex[i - 1];
        my.outindex[i - 1].incre = &my.outindex[i];
    }
    lwrap_mlocate((void **)&my.readindex, (sizeof(index_t) * my.buffboundry));
    my.readindex[0].val = 0;
    my.readindex[0].decre = &my.readindex[my.readbuffsize];
    my.readindex[my.readbuffsize].incre = &my.readindex[0];
    my.currreadindex = &my.readindex[0];
    for (int i = 1; i < my.buffboundry; i++) {
        my.readindex[i].val = i;
        my.readindex[i].decre = &my.readindex[i - 1];
        my.readindex[i - 1].incre = &my.readindex[i];
    }
    lwrap_mlocate((void **)&my.data, (sizeof(data_t) * my.databoundry));
    my.data[0].dataindex = 0;
    my.data[0].byteheader = 0;
    my.data[0].wbytecount = 0;
    my.data[0].wcolcount = 0;
    my.data[0].annotation = EOS;
    my.data[0].glyph = ASCII;
    my.data[0].val = EOS;
    my.data[0].before = &my.data[my.datasize];
    my.data[my.datasize].after = &my.data[0];
    for (int i = 1; i < my.databoundry; i++) {
        my.data[i].dataindex = i;
        my.data[i].wcolcount = 0;
        my.data[i].byteheader = 0;
        my.data[i].wbytecount = 0;
        my.data[i].annotation = EOS;
        my.data[i].glyph = ASCII;
        my.data[i].val = EOS;
        my.data[i].before = &my.data[i - 1];
        my.data[i - 1].after = &my.data[i];
    }
    my.currdata = &my.data[0];
    my.prevdata = my.currdata->before;
    my.nextdata = my.currdata->after;
    my.readdata = &my.data[0];
    my.outdata = &my.data[0];
    lwrap_mlocate((void **)&my.justify, (sizeof(justify_t) * my.justifyboundry));
    my.justify[0].remain = 0;
    my.justify[0].index = 0;
    my.justify[0].wcount = 0;
    my.justify[0].before = &my.justify[my.justifysize];
    my.justify[my.justifysize].after = &my.justify[0];
    for (int i = 1; i < my.justifyboundry; i++) {
        my.justify[i].remain = 0;
        my.justify[i].index = i;
        my.justify[i].wcount = 0;
        my.justify[i].before = &my.justify[i - 1];
        my.justify[i - 1].after = &my.justify[i];
    }
    my.currjustify = &my.justify[0];
    my.outjustify = &my.justify[0];
}

inline void alocate()
{
    acall();
    afring();
    aunicode();
    abuff();
}

inline void fjustify()
{
    if (!my.justifying) {
        my.writebuff[my.curroutindex->val] = SPACE;
        my.curroutindex = my.curroutindex->incre;
        return;
    }
    if (my.outjustify->remain == 0) {
        my.writebuff[my.curroutindex->val] = SPACE;
        my.curroutindex = my.curroutindex->incre;
        return;
    }
    if (my.outjustify->wcount < 2) {
        my.writebuff[my.curroutindex->val] = SPACE;
        my.curroutindex = my.curroutindex->incre;
        return;
    }
    if (my.outjustify->remain < 0 || my.outjustify->wcount < 2)
        return;
    int wseparator = my.outjustify->wcount - 1,
        spaces = my.outjustify->remain / wseparator,
        remainder = my.outjustify->remain % wseparator;
    if (remainder)
        spaces++;
    for (int i = 0; i < spaces; i++) {
        my.writebuff[my.curroutindex->val] = SPACE;
        my.curroutindex = my.curroutindex->incre;
    }
    my.writebuff[my.curroutindex->val] = SPACE;
    my.curroutindex = my.curroutindex->incre;
    my.outjustify->remain -= spaces;
    my.outjustify->wcount--;
}
